Add joystick interface

garaemon opened the pull request:
demo with teleop program:
https://www.youtube.com/watch?v=p_x-HRagLpo&feature=youtu.be

Add `geometry_msgs::PoseStamped` interface to move interactive marker from other ros nodes.
- `/rviz/moveit/plan`: same to `Plan` button
- `/rviz/moveit/execute`: same to `Execute` button
- `/rviz/moveit/select_planning_group`: a topic to select the planning group you want to control
- `/rviz/moveit/update_{goal,start}_state`: same to `<current> update` button
- `/rviz/moveit/$link_name`: a topic to specify the pose of the end effectors. the names of the topics will be like `/rviz/moveit/goal_base_footprint`, `/rviz/moveit/goal_l_wrist_roll_link` and so on.

See the sample movie:
http://youtu.be/sa9qyQLPGaQ

The sample code used to make the movie above.

``` python
#!/usr/bin/env python

import rospy

from std_msgs.msg import String, Empty
from geometry_msgs.msg import PoseStamped

rospy.init_node("moveit_visualization_sample")

# publisher to change the arm group
selector = rospy.Publisher("/rviz/moveit/select_planning_group", String)
plan = rospy.Publisher("/rviz/moveit/plan", Empty)
execute = rospy.Publisher("/rviz/moveit/execute", Empty)
reset_goal_state = rospy.Publisher("/rviz/moveit/update_goal_state", Empty)
# single arm 
rospy.sleep(5)
rospy.loginfo("move right_arm")
reset_goal_state.publish(Empty())
selector.publish(String("right_arm"))
arm_pub = rospy.Publisher("/rviz/moveit/move_marker/goal_r_wrist_roll_link", 
                          PoseStamped)
rospy.sleep(2)

goal = PoseStamped()
goal.header.frame_id = "/base_link"
goal.header.stamp = rospy.Time.now()
goal.pose.position.z = 1.1
goal.pose.position.x = 0.7
goal.pose.position.y = -0.3
goal.pose.orientation.w = 1.0
arm_pub.publish(goal)
rospy.sleep(1)
plan.publish(Empty())

rospy.sleep(5)

# dual arm
rospy.loginfo("move arms")
reset_goal_state.publish(Empty())
selector.publish(String("arms"))
larm_pub = rospy.Publisher("/rviz/moveit/move_marker/goal_l_wrist_roll_link", 
                           PoseStamped)
rarm_pub = rospy.Publisher("/rviz/moveit/move_marker/goal_r_wrist_roll_link", 
                           PoseStamped)
rospy.sleep(2)

rgoal = PoseStamped()
rgoal.header.frame_id = "/base_link"
rgoal.header.stamp = rospy.Time.now()
rgoal.pose.position.z = 1.1
rgoal.pose.position.x = 0.7
rgoal.pose.position.y = -0.3
rgoal.pose.orientation.w = 1.0
rarm_pub.publish(rgoal)

lgoal = PoseStamped()
lgoal.header.frame_id = "/base_link"
lgoal.header.stamp = rospy.Time.now()
lgoal.pose.position.z = 1.1
lgoal.pose.position.x = 0.7
lgoal.pose.position.y = 0.3
lgoal.pose.orientation.w = 1.0
larm_pub.publish(lgoal)

rospy.sleep(1)
plan.publish(Empty())
rospy.sleep(5)

# base
rospy.loginfo("move base")
reset_goal_state.publish(Empty())
selector.publish(String("base"))

base_pub = rospy.Publisher("/rviz/moveit/move_marker/goal_base_footprint", 
                           PoseStamped)
rospy.sleep(2)
goal = PoseStamped()
goal.header.frame_id = "/base_link"
goal.header.stamp = rospy.Time.now()
goal.pose.position.z = 0
goal.pose.position.x = 0.5
goal.pose.position.y = 0
goal.pose.orientation.w = 1.0
base_pub.publish(goal)
rospy.sleep(2)
plan.publish(Empty())
rospy.sleep(5)

```

garaemon committed:
adding PoseStamped topic to move the interactive marker from other ros nodes
such as joystick programs.

julienking committed:
returning int values

Commit:
Added field next to planning_time for planning_attempts
Now, ParallelPlanner terminates either due to timeout, or due to this many attempts.
Note, that ParallelPlanner run's Dijkstra's on all the nodes of all the sucessful plans (hybridize==true).

garaemon commented:
added new demo movie: https://www.youtube.com/watch?v=p_x-HRagLpo&feature=youtu.be

Commit:
moved planning_attempts down one row in gui to maintain gui width

sachinchitta committed:
0.5.19

sachinchitta committed:
0.5.19

arjungm committed:
Removed PlanningContext clear before planning call

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -66,7 +68,12 @@ RobotInteraction::RobotInteraction(const robot_model::RobotModelConstPtr &robot_
+  //   "/rviz/moveit/move_marker",
+  //   1,
+  //   &RobotInteraction::moveInteractiveMarker, this);
+  
```
Can you cleanup these comments?

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -671,6 +695,27 @@ bool RobotInteraction::updateState(
+  {
+    boost::unique_lock<boost::mutex> ulock(marker_access_lock_);
+    std::map<std::string, std::size_t>::const_iterator it = shown_markers_.find(name);
+    if (it != shown_markers_.end()) {
```
MoveIt style puts `{` on new line, thanks!

davetcoleman commented on visualization/motion_planning_rviz_plugin/src/motion_planning_display.cpp:
```diff
@@ -295,6 +297,17 @@ void MotionPlanningDisplay::onInitialize()
+{
+  if (!getRobotModel() || !robot_interaction_)
+    return;
+  if (getRobotModel()->hasJointModelGroup(msg->data)) {
```
New line

davetcoleman commented on visualization/motion_planning_rviz_plugin/src/motion_planning_frame.cpp:
```diff
@@ -176,6 +176,14 @@ MotionPlanningFrame::MotionPlanningFrame(MotionPlanningDisplay *pdisplay, rviz::
+  update_start_state_subscriber_ = nh.subscribe("/rviz/moveit/update_start_state",1,
+                                                &MotionPlanningFrame::updateStartStateCallback, this);
+  update_goal_state_subscriber_ = nh.subscribe("/rviz/moveit/update_goal_state",1,
+                                               &MotionPlanningFrame::updateGoalStateCallback, this);
```
Are all these subscribers always going to run, even when no joystick is being used? Seems like a little overhead

garaemon replied:
In order to reduce these subscription, One idea is adding another checkbox like "Allow remote control" to toggle this features. What do you think about this?

davetcoleman commented on visualization/motion_planning_rviz_plugin/src/motion_planning_frame_planning.cpp:
```diff
@@ -328,4 +328,39 @@ void MotionPlanningFrame::configureForPlanning()
   configureWorkspace();
 }
 
+void MotionPlanningFrame::planCallback(const std_msgs::EmptyConstPtr& msg)
```
Can we prepend these functions with something like `remotePlanCallback` to help others understand their purpose is for use with ROS messages / joysticks?

davetcoleman commented:
This is an exciting feature. I'll test this for you if you can lend me an extra joystick

garaemon commented:
my joy stick demo program (https://www.youtube.com/watch?v=p_x-HRagLpo&feature=youtu.be) requires a lot of extra-codes such as [jsk_teleop_joy](https://github.com/jsk-ros-pkg/jsk_control), [jsk_rviz_plugins](https://github.com/jsk-ros-pkg/jsk_visualization) and so on.

I will propose simpler code to control moveit rviz plugin from joystick.

andreasBihlmaier committed:
remove leading '/' from DEFAULT_PLANNING_SCENE_SERVICE; otherwise planning_scene_rviz_plugin fails with 'Failed to call service /get_planning_scene ...' if move_group is in a namespace different from '/'

andreasBihlmaier committed:
service must have access to global namespace

jbohren committed:
motion_planning_rviz_plugin: add move_group namespace option

This allows multiple motion_planning_rviz_plugin /
planning_scene_rviz_plugin to be used in RViz and connect to
differently-namespaced move_group nodes.

davetcoleman committed:
Made loading octomap_monitor optional in planning_scene_monitor when using WorldGeometryMonitor

davetcoleman commented:
@garaemon did you mention you've made the changes to the example node necessary to test?

garaemon committed:
add checkbox to toggle if moveit rviz plugin subscribes
the topics to be used for communication to the external ros nodes.

update moveit_joy.py to parse srdf to know planning_groups and the
names of the end effectors and support multi-endeffector planning groups.

garaemon commented:
I don't finish thei PR yet, but today I resume working on this issue

2014年8月4日月曜日、Dave Colemannotifications@github.comさんは書きました:

> @garaemon https://github.com/garaemon did you mention you've made the
> changes to the example node necessary to test?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-planning/moveit_ros/pull/471#issuecomment-51018088
> .

## 

from iPhone

garaemon commented:
I added a python script and launch file to run joystick demo.
https://www.youtube.com/watch?v=oDFbLylrhig

And, added new check box to toggle communication with external ros nodes.
![screenshot_from_2014-08-05 12 43 48](https://cloud.githubusercontent.com/assets/40454/3808046/8c5227fe-1c71-11e4-88ec-a8bd74e0b3e2.png)

However I don't know where those scripts should be put.
Now the scripts are located under python and joystick_control directory.

``` sh
# Plug your xbox or ps3 joystick to your computer.
$ roslaunch pr2_moveit_config demo.launch
$ roslaunch moveit_ros_visualization pr2_sample.launch
```

cheat sheet

```
left analog stick: +-x/y
L2/R2: +-z
L1/R1: +-yaw
left/right: +-roll
up/down: +-pitch
triangle/cross: change planning group
square: plan
circle: execute
```

davetcoleman commented:
Looking great. I look forward to the tweaks we discussed. 

Also, perhaps the checkbox could just say "Allow External Communication" or "Allow External Control" to prevent it from possibly stretching that QT box.

garaemon commented:
https://www.youtube.com/watch?v=wbl90WPAUJg

I updated the code:
summary:
- add new checkbox "Allow External Communication" to toggle topics to communicate with other ROS nodes
- parse srdf file directly from the python joystick script. No need to specify planning_groups.
- support multi end effectors. (use start/select to choose the end effector)

davetcoleman commented on visualization/package.xml:
```diff
@@ -29,7 +29,7 @@
+  <build_depend>rospy</build_depend>
   <run_depend>roscpp</run_depend>
   <run_depend>rviz</run_depend>
   <run_depend>moveit_ros_planning_interface</run_depend>
```
technically you need to add a run_depend for `joy` since the python code uses it.

however, im not sure if we want moveit_ros to depend on something that is slightly hardware dependent. @isucan ? part of me wants to suggest the python script and launch file should go in a separate package, like moveit_joystick_control, but that would be extra work to maintain at the same time.

davetcoleman commented:
I started documenting how to use this, please fill in more details as you see fit. This file should probably go in `moveit_ros/visualization/doc/joystick.rst`:

```
Joystick Control of Rviz Interactive Markers
@@@@@@@@@@@@@@@@@@@@@@@@@@@@

Run
==================
roslaunch moveit_ros_visualization joy_sample.launch

Joystick Cheat Sheet
==================

left analog stick: +-x/y
L2/R2: +-z
L1/R1: +-yaw
left/right: +-roll
up/down: +-pitch
triangle/cross: change planning group
square: plan
circle: execute

```

Can you add an updated version of this to the PR?

davetcoleman commented on visualization/motion_planning_rviz_plugin/src/ui/motion_planning_rviz_plugin_frame.ui:
```diff
@@ -639,6 +639,16 @@
           <item>
+           <widget class="QCheckBox" name="allow_external_program">
+            <property name="text">
+             <string>Allow External Communication</string>
```
I think this is still causing the GUI box to expand more than it should. Can you shorten it to just "Allow External Comm."?

davetcoleman commented on visualization/joystick_control/launch/joy_sample.launch:
```diff
@@ -0,0 +1,13 @@
+<launch>
```
Perhaps name this file just "joystick_control.launch"? It is no longer really a sample but a full feature node, right?

davetcoleman commented:
I tried running this just now but am missing something. Could you show me how to use it tomorrow?

Thanks for this work!

davetcoleman committed:
Documentation

garaemon committed:
Update joystick documentation and rename the
the launch file for joy stick program.
Shorten the message the check box to toggle
communication with joy stick script.

davetcoleman committed:
Python wrapper for getEndEffectorTips()

davetcoleman committed:
Added wrapper for new function

garaemon committed:
Remove sample launch file for joystick and update
joystick python script.
1) Use moveit-python binding to parse SRDF.
2) Make the speed slower to control the marker from joystick.
3) Change joystick button mapping to be suitable for the users.

garaemon commented:
updated the joystick script to initialize the poses of the markers correctly.

davetcoleman committed:
Better user output, kinematic solver error handling, disclaimer

davetcoleman commented:
garaemon, thanks, this is really looking shaping up as a great new feature!

I did some more testing today and made a number of little changes: https://github.com/garaemon/moveit_ros/pull/2

Other feedback I have not addressed yet:
- Can x/y be a little slower, e.g. less sensitive? maybe half the speed?
- When interactive marker goes beyond valid position, can it bounce back somehow? maybe after a timeout? or stop?
- What do you think about switching the buttons for the functionality of changing planning groups with the functionality of changing end effectors? It seems much more common to want to change end effectors i.e. with a humanoid that it will be to change planning groups

Note: I also changed the Setup Assistant launch file to use `dev` as argument instead of `DEV` to match convention.

Thanks!

isucan commented:
This looks neat!
What do you think @acornacorn ?

garaemon commented:
> Can x/y be a little slower, e.g. less sensitive? maybe half the speed?

make the speed slower at https://github.com/garaemon/moveit_ros/commit/75808c2770c3836249ab659264deaa048f1e1f96

> When interactive marker goes beyond valid position, can it bounce back somehow? maybe after a timeout? or stop?

Hmm, it may mean that the interactive markers cannot pass the obstacles. And it requires a lot of modification about communication between rviz plugin and joy stick...

> What do you think about switching the buttons for the functionality of changing planning groups with the functionality of changing end effectors? It seems much more common to want to change end effectors i.e. with a humanoid that it will be to change planning groups

I swapped the functionalities at https://github.com/garaemon/moveit_ros/commit/1cd77322531ed825def759330fc62902652dc27d

garaemon committed:
update joystick documentation according to the latest implementation

davetcoleman commented:
> Hmm, it may mean that the interactive markers cannot pass the obstacles. And it requires a lot of modification about communication between rviz plugin and joy stick...

That is fine, I agree it would be difficult and perhaps future work.

davetcoleman commented:
Thanks for all the work so far! I did some more testing today and everything looks great on HRP2 with 2 arms and 2 legs planning at the same time using KDL IK solver.

Two remaining issues:
- The Rviz Motion Planning plugin bottom left box seems to always be refreshing with and showing a blinking progress bar indicator. Maybe this should just be ignored, at least for now, but its not perfect. Do you know of a quick fix? @acornacorn would probably know best about this feature.
- Could you rebase this PR, at least the first chunk before my commit? Currently this PR is 30 commits long.

Thanks again!

garaemon commented:
Rebased this branch.

davetcoleman commented on robot_interaction/include/moveit/robot_interaction/robot_interaction.h:
```diff
@@ -228,6 +239,14 @@ class RobotInteraction
+  std::vector<ros::Subscriber> int_marker_move_subscribers_;
+  // the array of the names of the topics which need to be subscribed
+  // to move the interactive markers by other ROS nodes
+  std::vector<std::string> int_maker_move_topics_;
```
I believe you meant `int_marker_move_topics_`

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -561,6 +572,42 @@ void RobotInteraction::addInteractiveMarkers(
+    boost::unique_lock<boost::mutex> ulock(marker_access_lock_);
+    if (int_marker_move_subscribers_.size() != 0) {
+      
+    }
```
Remove this empty bracket?

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -561,6 +572,42 @@ void RobotInteraction::addInteractiveMarkers(
+    }
+    else {
+      ros::NodeHandle nh;
+      for (size_t i = 0; i < int_maker_move_topics_.size(); i++) {
```
MoveIt! format has brackets on separate line :-)

garaemon replied:
Hmm, I don't know... 
Anyway, I fix that

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -671,6 +718,34 @@ bool RobotInteraction::updateState(
+    }
+  }
+  
+//  }
```
Cleanup comment

davetcoleman commented on robot_interaction/src/robot_interaction.cpp:
```diff
@@ -671,6 +718,34 @@ bool RobotInteraction::updateState(
 
+void RobotInteraction::moveInteractiveMarker(const std::string name, const geometry_msgs::PoseStampedConstPtr& msg)
+{
+  {
```
Why extra brackets here?

davetcoleman commented:
It looks like the rebase did some weird copying of commits, but I did a local `git diff hydro-devel add-joystick-node` and this PR doesn't change anything except joystick related stuff. So I think the noise is ok.

I also did a little deeper code review, sorry for the last minute comments!

davetcoleman commented:
@garaemon  can you address my final round of comments so we can finish this?

garaemon commented:
sorry, I missed the last comment. I will fix it today.

2014年9月5日金曜日、Dave Colemannotifications@github.comさんは書きました:

> @garaemon https://github.com/garaemon can you address my final round of
> comments so we can finish this?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-planning/moveit_ros/pull/471#issuecomment-54565735
> .

## 

from iPhone

garaemon committed:
Fix coding style according to the moveit style

garaemon commented:
I fixed coding style around if-else brackets

davetcoleman commented:
Thanks for all the work @garaemon 

Now we need to trigger the moveit website to display our documentation - I'm not sure how to do this but will work on it.

davetcoleman commented:
Ok, once moveit_ros is released the doxygen will appear in [this directory](http://docs.ros.org/hydro/api/moveit_ros_visualization/html/doc/)

Then we add the link to http://moveit.ros.org/documentation/tutorials/

I think you could go ahead and announce the feature on the mailing list.

garaemon commented:
Thank you for your support!

2014年9月5日金曜日、Dave Colemannotifications@github.comさんは書きました:

> Ok, once moveit_ros is released the doxygen will appear in this directory
> http://docs.ros.org/hydro/api/moveit_ros_visualization/html/doc/
> 
> Then we add the link to http://moveit.ros.org/documentation/tutorials/
> 
> I think you could go ahead and announce the feature on the mailing list.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ros-planning/moveit_ros/pull/471#issuecomment-54581696
> .

## 

from iPhone
